snapfaas/src/sched/messages.proto

```diff
@@ -0,0 +1,61 @@
+syntax = "proto3";
+
+package snapfaas.sched.messages;
+
+// Begin of request messages
+message GetJob {
+    uint64 id = 1;
+}
+
+message FinishJob {
+    string id     = 1;
+    bytes  result = 2;
+}
+
+message Invoke {
+    bytes invoke = 1;
+}
+
+message UpdateResource {
+    bytes info = 1;
+}
+
+message DropResource {}
+message ShutdownAll {}
+// End of request messages
+
+// Begin of response messages
+message ProcessJob {
+    string id     = 1;
+    bytes  invoke = 2;
+}
+
+message Shutdown {}
```

Comment:
What's the meaning of each of these messages?

Reply:
`UpdateResource` is used for the worker machine to update its local resource info (i.e, cached VM) because the global resource manager doesn't evict any VM explicitly so only the local manager can tell.

`DropResource` is used for the worker machine to grateful shutdown itself. It tells the scheduler the worker is no longer available.

`ShutdownAll` is used in the `ctrl-c-handler` in `multivm`, indicating the scheduler to shut down all workers. After receiving such a message, the scheduler sends `Shutdown` to all workers, and workers shut down themselves when getting `Shutdown`. Without it, `multivm` hangs when pressing C-c.

To get and process a request, a worker does the following:
1. Send `GetJob` to tell the scheduler it is ready
2. Receive `ProcessJob` from the scheduler (`invoke` contains `LabeledInvoke`)
3. Send `FinishJob` to return the result to the scheduler (for synchronous invocation)